tibanna_4dn/pony_utils.py

```diff
@@ -386,159 +395,761 @@ def is_prod():
-        keys = []
-        for v in self.all_files():
-            if argname == v.argument_name:
-                keys.append(v.key)
-        return keys
+    def format_if_extras(self, argname):
+        format_if_extras = []
+        for v in self.ff_files:
+            if argname == v['workflow_argument_name'] and 'format_if_extra' in v:
+                format_if_extras.append(v['format_if_extra'])
+        return format_if_extras
+
+    def status(self, argname=None, pf_uuid=None):
+        """check file existence as status either per argname or per pf object"""
+        if argname:
+            key = self.file_key(argname)
+        elif pf_uuid:
+            key = self.file_key(pf_uuid=pf_uuid)
+            argname = self.pf2argname(pf_uuid)
+        else:
+            raise Exception("Either argname or pf_uuid must be provided to get status.")
+        if not isinstance(key, list):
+            key = [key]
+        for k in key:
+            try:
+                exists = self.s3(argname).does_key_exist(k, self.bucket(argname))
+                if not exists:
+                    return "FAILED"
+            except:
+                return "FAILED"
+        return "COMPLETED"
+
+    def genome_assembly(self, pf_uuid):
+        if hasattr(self.pf(pf_uuid), 'genome_assembly'):
+            return self.pf(pf_uuid).genome_assembly
+        else:
+            return None
+
+    # update functions for PF
+    def update_all_pfs(self):
+        if self.pf_output_files:
+            for pf_uuid in self.pf_output_files:
+                self.update_pf(pf_uuid)
+
+    def update_pf(self, pf_uuid):
+        if self.status(pf_uuid=pf_uuid) == 'COMPLETED':
+            self.add_updates_to_pf(pf_uuid)
+            self.add_updates_to_pf_extra(pf_uuid)
+            self.add_higlass_to_pf(pf_uuid)
+
+    def add_updates_to_pf(self, pf_uuid):
+        """update md5sum, file_size, status for pf itself"""
+        # update the class object
+        self.pf(pf_uuid).md5sum = self.md5sum(pf_uuid=pf_uuid)
+        self.pf(pf_uuid).file_size = self.filesize(pf_uuid=pf_uuid)
+        self.pf(pf_uuid).status = 'uploaded'
+        # prepare for fourfront patch
+        self.add_to_pf_patch_items(pf_uuid, ['md5sum', 'file_size', 'status'])
+
+    def add_updates_to_pf_extra(self, pf_uuid):
+        """update md5sum, file_size, status for extra file"""
+        for extra in self.all_extra_files(pf_uuid=pf_uuid):
+            pf_extra = self.pf_extra_file(pf_uuid, extra['file_format'])
+            extra_key = extra['upload_key']
+            # update the class object
+            pf_extra['md5sum'] = self.md5sum(pf_uuid=pf_uuid, secondary_key=extra_key)
+            pf_extra['file_size'] = self.filesize(pf_uuid=pf_uuid, secondary_key=extra_key)
+            pf_extra['status'] = 'uploaded'
+        # prepare for fourfront patch
+        self.add_to_pf_patch_items(pf_uuid, 'extra_files')
+
+    def add_higlass_to_pf(self, pf_uuid):
```

Comment:
Maybe`hgcf` is never set for files without `extra_files`? An example would be bigwig (bw) files

Reply:
https://github.com/4dn-dcic/tibanna/blob/69892fb8477a6e0ffce85c5b216b06a7526b3675/tibanna_4dn/pony_utils.py#L883

Reply:
Okay, I see how it works. Was going a bit too fast when I made that comment!